### DISCUSSION TITLE
Support multi-chain wallets in a nicer way

### DISCUSSION THREAD

**Discussion by ethernet8023** [wevm/wagmi#396] [Ideas]
Hey! I've been using wagmi + RainbowKit for an upgrade to [Vaportrade](https://vaportrade.net/)
Vaportrade's primary userbase is users of the [Sequence](https://sequence.app) wallet, which supports being connected to multiple chains at once.
I'd like to propose that wagmi support multi-chain wallets better, such that devs can request a signer or a provider for a specific chain without setting "the current chain" of the wallet.
## Motivation

To provide a seamless integration with multichain wallets, devs need to be able to access things like a `Signer` without requesting that a wallet change its "current chain", since not all multichain wallets care about a "current chain".
## Prior Art and API inspiration
Vaportrade needs this functionality, and so I've built a small version myself.
In my current version, I manually type-cast a connector to `(Connector | (Connector & MultiChainConnector))`, so if `isMultiChain` is true, I can access the getProvider & getSigner's additional arguments.
```ts
interface MultiChainConnector<Provider = any> {
  getProvider(create?: boolean, chainId?: number): Provider;
  getSigner(chainId?: number): Promise<Signer>;
  isMultiChain: true;
}
```

If this were to be integrated into `wagmi` I would imagine a MultiChainConnector subclass that multichain wallets could `extend`, and we could get rid of `isMultiChain` in favor of an `instanceof` call that would show you the extra `chainId` parameters.
```ts
abstract class MultiChainConnector<Provider = any, Options = any> extends Connector<Provider, Options> {
  abstract getProvider(create?: boolean, chainId?: number): Provider;
  abstract getSigner(chainId?: number): Promise<Signer>;
}
```

and it would be used like
```js
const chainId = ...;
let signer;
if (connector instanceof MultiChainConnector) {
  signer = await connector.getSigner(chainId);
} else {
  const { data, error } = await switchNetwork(chainId);
  if(error) {
    throw error;
  }
  if (data.chain.id === chainId) {
    throw new Error("Failed to switch network");  
  }
  signer = connector.getSigner();
}
```
The `switchChain` method could become a no-op for `MultiChainConnector` instances.

## What else?
Any thoughts? Would love to upstream something like this :)

**Comment by jmrossy** [comment]
I'm also curious about how libs like wagmi can better support multi-chain apps.

As far as I can tell, all popular libs are built around the assumption that they're connected to one chain at a time. As bridging solutions become more mature, it'll be more common for apps to need live connections to multiple chains simultaneously. 

Your proposal for a `MultiChainConnector` can work but having its methods not match the `Connector` methods' signatures is maybe not ideal.

**Comment by Ding-Fan** [comment]
I'm using https://docs.family.co/connectkit , looks good to me

**Reply by kombos** [reply]
> I'm using [docs.family.co/connectkit](https://docs.family.co/connectkit) , looks good to me

do you mean that multichain connectedness works with ConnectKit ? how do we perform write operations (send transactions) to multiple chains without the need to switch network , in connectKit ?
